General concepts of working with Orbit, Ephem, and Phys objects
===============================================================

The classes Orbit, Ephem, and Phys are all derived from the same base class ([DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass)), meaning that they can all be used in the same way. 

The core of [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) is an astropy [QTable](http://docs.astropy.org/en/stable/api/astropy.table.QTable.html#astropy.table.QTable) object, which is an astropy [Table](http://docs.astropy.org/en/stable/table/) that is aware of [units](http://docs.astropy.org/en/stable/units/). [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) provides a number of convenience functions that make it easy to create and modify the underlying [QTable](http://docs.astropy.org/en/stable/api/astropy.table.QTable.html#astropy.table.QTable) object. If these convenience functions do not cover what you are trying to do, you can directly address and modify the [QTable](http://docs.astropy.org/en/stable/api/astropy.table.QTable.html#astropy.table.QTable) object.

For some introductional examples on how to create, access, and modify [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) objects, please have a look at the [documentation](https://sbpy.readthedocs.io/en/latest/sbpy/data.html).

In the following sections, we provide some examples that are close to real use cases.

Adding a column and converting units
---------------------------

We obtain ephemerides for our target asteroid from JPL Horizons.

In [1]:
from astropy.time import Time
from sbpy.data import Ephem

epoch1 = Time('2018-09-01 12:00', scale='utc')
epoch2 = Time('2018-09-30 12:00', scale='utc')
eph = Ephem.from_horizons('2018 RE3',
                          location='568',
                          epochs={'start': epoch1,
                                  'stop': epoch2,
                                  'step': '5d'},
                          skip_daylight=True)
eph.table

<QTable masked=True length=6>
targetname    datetime_str   datetime_jd    H    ...  PABLon   PABLat  timescale
                                  d        mag   ...   deg      deg             
  str10          str17         float64   float64 ... float64  float64     str3  
---------- ----------------- ----------- ------- ... -------- -------- ---------
(2018 RE3) 2018-Sep-01 12:00   2458363.0    27.3 ... 338.9039   5.2605       UTC
(2018 RE3) 2018-Sep-06 12:00   2458368.0    27.3 ... 340.6555   6.4863       UTC
(2018 RE3) 2018-Sep-11 12:00   2458373.0    27.3 ... 340.0626  13.4255       UTC
(2018 RE3) 2018-Sep-16 12:00   2458378.0    27.3 ...  76.7926  43.5756       UTC
(2018 RE3) 2018-Sep-21 12:00   2458383.0    27.3 ...  80.6791   -7.097       UTC
(2018 RE3) 2018-Sep-26 12:00   2458388.0    27.3 ...  83.0692 -10.5497       UTC

Sky motion rates are provided for both RA and Dec in units of arcsec per hour:

In [2]:
print(eph['RA_rate', 'DEC_rate'])

 RA_rate    DEC_rate 
arcsec / h arcsec / h
---------- ----------
  -62.6556   36.73963
  -144.843   97.00795
  -1184.37   952.1053
  -429.792   -484.058
   -39.095   -72.7217
  0.090585   -28.8419


Let's assume that we are interested in the absolute sky motion in units of arcsec per second. It would be handy to add a column to `eph` that contains this quantity. This can be easily done using the [add_column](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass.add_column) function. Finally, we want to print a table containing epoch, the distance to the observer, and the corresponding absolute sky motion:

In [3]:
import numpy as np
eph.add_column(np.sqrt(eph['RA_rate']**2 + eph['DEC_rate']**2).to('arcsec/second'), name='abs_rate')
eph['datetime_str', 'delta', 'abs_rate']

<QTable masked=True length=6>
   datetime_str        delta             abs_rate      
                         AU             arcsec / s     
      str17           float64            float64       
----------------- ---------------- --------------------
2018-Sep-01 12:00 0.07201126545981 0.020175779667146194
2018-Sep-06 12:00 0.04189463679787  0.04842427349261143
2018-Sep-11 12:00 0.01285453701319   0.4221159222171138
2018-Sep-16 12:00 0.01819643837712   0.1798132842091663
2018-Sep-21 12:00 0.04756413980489 0.022934529529610993
2018-Sep-26 12:00 0.07789742321353 0.008011678403266262

The function [add_column](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass.add_column) takes two parameters: the data in form of a list or an array, and the name of the new column. The data that we provide here consists of two parts. We derive the absolute sky motion as the geometric sum of the RA and Dec rates (`np.sqrt(eph['RA_rate']**2 + eph['DEC_rate']**2)`). Remember that each of these rates is in units of arcsec per hour. In order to convert the rates to arcsec per sec, we have to apply `.to('arcsec/second')` to this array. Finally, we assign a name to this new column.

Filtering table content
-----------------------

We have a list of 5 asteroids that we would like to observe on the night of 2018-09-12 at the Discovery Channel telescope. However, we don't want to observe these asteroids when they have an absolute sky motion rate faster than 0.1 arcsec/s and when the Moon is up.

In [4]:
import numpy as np
import astropy.units as u
from astropy.time import Time
from sbpy.data import Ephem

# target list
targets = ['2018 RR4', '2018 RE3', '2018 RC4', '2018 RQ2', '2018 RC1']

epoch1 = Time('2018-09-13 00:00', scale='utc')
epoch2 = Time('2018-09-14 00:00', scale='utc')
eph = Ephem.from_horizons(targets,
                          location='G37',
                          epochs={'start': epoch1,
                                  'stop': epoch2,
                                  'step': '10m'},
                          skip_daylight=True)
print(len(eph.table))

345


A total of 345 ephemerides have been queried for the 5 asteroids. Now we apply the absolute sky motion rate (see example above) and moon filters; if the moon is up, `eph['flags']` will be set to `'m'` (see https://ssd.jpl.nasa.gov/?horizons_doc&table_quantities#table_quantities):

In [5]:
eph = eph[np.sqrt(eph['RA_rate']**2 + eph['DEC_rate']**2).to('arcsec/second') < 0.1*u.arcsec/u.second]
eph = eph[eph['flags'] != 'm'] 

eph

<QTable masked=True length=55>
targetname    datetime_str      datetime_jd    ...  PABLon   PABLat timescale
                                     d         ...   deg      deg            
  str10          str17            float64      ... float64  float64    str3  
---------- ----------------- ----------------- ... -------- ------- ---------
(2018 RC1) 2018-Sep-13 04:00 2458374.666666667 ... 350.8126  4.4012       UTC
(2018 RC1) 2018-Sep-13 04:10 2458374.673611111 ... 350.8204  4.3984       UTC
(2018 RC1) 2018-Sep-13 04:20 2458374.680555556 ... 350.8281  4.3956       UTC
(2018 RC1) 2018-Sep-13 04:30      2458374.6875 ... 350.8357  4.3928       UTC
(2018 RC1) 2018-Sep-13 04:40 2458374.694444444 ... 350.8434  4.3899       UTC
(2018 RC1) 2018-Sep-13 04:50 2458374.701388889 ...  350.851  4.3871       UTC
(2018 RC1) 2018-Sep-13 05:00 2458374.708333333 ... 350.8586  4.3844       UTC
(2018 RC1) 2018-Sep-13 05:10 2458374.715277778 ... 350.8662  4.3816       UTC
(2018 RC1) 2018-Sep-13 05:20 2458374.722222222 ... 350.8737  4.3788       UTC
       ...               ...               ... ...      ...     ...       ...
(2018 RC1) 2018-Sep-13 11:30 2458374.979166667 ... 351.1526  4.2735       UTC
(2018 RC1) 2018-Sep-13 11:40 2458374.986111111 ... 351.1605  4.2704       UTC
(2018 RC1) 2018-Sep-13 11:50 2458374.993055556 ... 351.1685  4.2673       UTC
(2018 RC1) 2018-Sep-13 12:00         2458375.0 ... 351.1765  4.2642       UTC
(2018 RC1) 2018-Sep-13 12:10 2458375.006944444 ... 351.1845  4.2611       UTC
(2018 RC1) 2018-Sep-13 12:20 2458375.013888889 ... 351.1926  4.2579       UTC
(2018 RC1) 2018-Sep-13 12:30 2458375.020833333 ... 351.2007  4.2548       UTC
(2018 RC1) 2018-Sep-13 12:40 2458375.027777778 ... 351.2089  4.2516       UTC
(2018 RC1) 2018-Sep-13 12:50 2458375.034722222 ... 351.2171  4.2484       UTC
(2018 RC1) 2018-Sep-13 13:00 2458375.041666667 ... 351.2253  4.2452       UTC

As it turns out, only asteroid 2018 RC1 is observable under these conditions.